### PR TITLE
feat(dashmate): core log file and debug categories

### PR DESCRIPTION
**Pull request by pshenmic**
## Issue being fixed or feature implemented

Change dashcore logging from stdout to external file and put it in the logs folder in dashmate configuration directory.

## What was done?

* changed printtoconsole to 0
* added new dashcore core.log log file in ~/.dashmate/logs/<network_name>

## How Has This Been Tested?

Locally

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/dashmate/configs/schema/configJsonSchema.js [hunk 1]
```diff
@@ -243,9 +243,29 @@ module.exports = {
...
-        debug: {
-          type: 'integer',
-          enum: [0, 1],
+        log: {
+          type: 'object',
+          properties: {
+            file: {
+              type: 'object',
+              properties: {
+                level: {
```
Comment (shumkov) [inline packages/dashmate/configs/schema/configJsonSchema.js:252]: Let's call it `categories` and it must be `enum` with `all` and other categories

File: packages/dashmate/configs/system/base.js [hunk 2]
```diff
@@ -79,7 +79,12 @@ module.exports = {
       minimumDifficultyBlocks: 0,
       powTargetSpacing: 150,
     },
-    debug: 0,
+    log: {
+      file: {
+        level: ['0'],
```
Comment (shumkov) [inline packages/dashmate/configs/system/base.js:84]: Just empty categories

File: packages/dashmate/docker-compose.yml [hunk 3]
```diff
@@ -10,6 +10,7 @@ services:
     volumes:
       - core_data:/home/dash
       - ${DASHMATE_HOME_DIR:?err}/${CONFIG_NAME:?err}/core/dash.conf:/home/dash/.dashcore/dash.conf
+      - ${CORE_LOG_FILE_PATH:?err}:/tmp/debug.log
```
Comment (shumkov) [inline packages/dashmate/docker-compose.yml:13]: I think we need to do in the same way is we did with Drive

File: packages/dashmate/src/listr/tasks/setup/setupLocalPresetTaskFactory.js [hunk 4]
```diff
@@ -172,8 +172,11 @@ function setupLocalPresetTaskFactory(
                 config.set('docker.network.subnet', `172.24.${nodeIndex}.0/24`);
 
                 // Setup Core debug logs
+                const coreLogFilePath = path.join(HOME_DIR_PATH, 'logs', config.getName(), 'core.log');
+                config.set('core.log.file.path', coreLogFilePath);
+
                 if (ctx.debugLogs) {
-                  config.set('core.debug', 1);
+                  config.set('core.log.file.level', ['1']);
```
Comment (shumkov) [inline packages/dashmate/src/listr/tasks/setup/setupLocalPresetTaskFactory.js:179]: `['all']`

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/configs/schema/configJsonSchema.js [hunk 5]
```diff
@@ -243,9 +243,34 @@ module.exports = {
...
-          type: 'integer',
-          enum: [0, 1],
+        log: {
+          type: 'object',
+          properties: {
+            file: {
+              type: 'object',
+              properties: {
+                categories: {
+                  type: 'array',
```
Comment (shumkov) [inline packages/dashmate/configs/schema/configJsonSchema.js:253]: uniqueItems: true

File: packages/dashmate/src/core/startCoreFactory.js [hunk 6]
```diff
@@ -57,6 +59,21 @@ function startCoreFactory(
...
+    if (fs.existsSync(logFilePath) && fs.lstatSync(logFilePath).isDirectory()) {
+      fs.rmSync(logFilePath, { recursive: true });
+    }
+
+    if (!fs.existsSync(logFilePath)) {
+      fs.mkdirSync(path.dirname(logFilePath), { recursive: true });
+      fs.writeFileSync(logFilePath, '');
+    }
+
+    console.log(config.toEnvs());
```
Comment (shumkov) [inline packages/dashmate/src/core/startCoreFactory.js:74]: Remove console logs

File: packages/dashmate/src/listr/tasks/startNodeTaskFactory.js [hunk 7]
```diff
@@ -6,6 +6,18 @@ const { Observable } = require('rxjs');
 
 const { NETWORK_LOCAL } = require('../../constants');
 
+const ensureLogFilePath = (logFilePath) => {
```
Comment (shumkov) [inline packages/dashmate/src/listr/tasks/startNodeTaskFactory.js:9]: I guess you want to create a function in `src/docker` and call it `ensureFileMountExists` so you can reuse it here and `startCoreFactory`

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/src/createDIContainer.js [hunk 8]
```diff
@@ -111,6 +112,7 @@ async function createDIContainer() {
     migrateConfigFile: asValue(migrateConfigFile),
     isHelper: asValue(process.env.DASHMATE_HELPER === '1'),
     getConnectionHost: asClass(getConnectionHostFactory).singleton(),
+    ensureFileMountExists: asClass(ensureFileMountExistsFactory).singleton(),
```
Comment (shumkov) [inline packages/dashmate/src/createDIContainer.js:115]: asFactory

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/src/createDIContainer.js (hunk 8 above)
Comment (pshenmic) [reply packages/dashmate/src/createDIContainer.js]: There is no `asFactory`, changed to `asFunction` instead

**Comment by pshenmic** [comment]
Test results:

#### Empty categories `[]`
* No log output in docker container stdout
* No debug log lines found in core.log

#### Enable few categories `["rpc", "http", "walletdb"]`
* No log output in docker container stdout
* `debug=<category>` is added for each one in dash.conf
* Debug category lines can be seen in core.log

#### Enable all debug categories `["all"]`
* No log output in docker container stdout
* `debug=1` is added in the dash.conf
* All debug logs can be found in core.log with all enabled categories

**Comment by pshenmic** [comment]
After discussion with @shumkov, we decided to completely remove printtoconsole option and get stdout logs in the docker container back.

**Review by shumkov — APPROVED** [review]
👍

**Pull request merged**